File: Tests/SPFNConformanceTests/FixtureLoading.swift
// SPFN Mobile — shared fixture loading for the Swift conformance suite.
//
// The Kotlin suite under android/spfn-auth/src/test/kotlin reads the SAME files from
// the SAME directory. Neither suite carries its own copy of an expected value, so a
// vector cannot drift on one platform without the other noticing.

import Foundation
import SPFNCore
import XCTest

enum Fixtures
{
    /// Repository root, located relative to this source file so the suite works from
    /// any working directory.
    static let repoRoot: URL = URL(fileURLWithPath: #filePath)
        .deletingLastPathComponent()   // Tests/SPFNConformanceTests
        .deletingLastPathComponent()   // Tests
        .deletingLastPathComponent()   // repo root

    static let directory: URL = repoRoot.appendingPathComponent("Contracts/fixtures")

    /// Loads a fixture through the SDK's own strict parser, so reading the evidence
    /// exercises the thing being tested.
    static func load(_ relativePath: String) throws -> SPFNCanonicalValue
    {
        let data = try Data(contentsOf: directory.appendingPathComponent(relativePath))
        return try SPFNCanonicalJSON.parse([UInt8](data))
    }
}

// Small readers so the test bodies stay about the contract rather than about optionals.
extension SPFNCanonicalValue
{
    func members(file: StaticString = #filePath, line: UInt = #line) throws -> [String: SPFNCanonicalValue]
    {
        guard case .object(let members) = self
        else
        {
            XCTFail("expected an object, got \(self)", file: file, line: line)
            throw ConformanceFailure.shape
        }
        return members
    }

    func elements(file: StaticString = #filePath, line: UInt = #line) throws -> [SPFNCanonicalValue]
    {
        guard case .array(let elements) = self
        else
        {
            XCTFail("expected an array, got \(self)", file: file, line: line)
            throw ConformanceFailure.shape
        }
        return elements
    }

    func text(file: StaticString = #filePath, line: UInt = #line) throws -> String
    {
        guard case .string(let value) = self
        else
        {
            XCTFail("expected a string, got \(self)", file: file, line: line)
            throw ConformanceFailure.shape
        }
        return value
    }

    func number(file: StaticString = #filePath, line: UInt = #line) throws -> Int64
    {
        guard case .integer(let value) = self
        else
        {
            XCTFail("expected an integer, got \(self)", file: file, line: line)
            throw ConformanceFailure.shape
        }
        return value
    }
}

extension Dictionary where Key == String, Value == SPFNCanonicalValue
{
    func text(_ key: String, file: StaticString = #filePath, line: UInt = #line) throws -> String
    {
        guard let value = self[key]
        else
        {
            XCTFail("fixture is missing '\(key)'", file: file, line: line)
            throw ConformanceFailure.shape
        }
        return try value.text(file: file, line: line)
    }

    func number(_ key: String, file: StaticString = #filePath, line: UInt = #line) throws -> Int64
    {
        guard let value = self[key]
        else
        {
            XCTFail("fixture is missing '\(key)'", file: file, line: line)
            throw ConformanceFailure.shape
        }
        return try value.number(file: file, line: line)
    }

    func list(_ key: String, file: StaticString = #filePath, line: UInt = #line) throws -> [SPFNCanonicalValue]
    {
        guard let value = self[key]
        else
        {
            XCTFail("fixture is missing '\(key)'", file: file, line: line)
            throw ConformanceFailure.shape
        }
        return try value.elements(file: file, line: line)
    }
}

enum ConformanceFailure: Error
{
    case shape
}

File: android/spfn-auth/src/test/kotlin/xyz/superfunction/spfn/conformance/FixtureLoading.kt
// SPFN Mobile — shared fixture loading for the Kotlin conformance suite.
//
// The Swift suite under Tests/SPFNConformanceTests reads the SAME files from the SAME
// directory. Neither suite carries its own copy of an expected value, so a vector
// cannot drift on one platform without the other noticing.

package xyz.superfunction.spfn.conformance

import xyz.superfunction.spfn.core.SpfnCanonicalJson
import xyz.superfunction.spfn.core.SpfnCanonicalValue
import java.io.File

object Fixtures
{
    /**
     * Repository root, injected by the Gradle test task rather than guessed, so the
     * suite reads the same checkout the build is running against.
     */
    val repoRoot: File = File(
        requireNotNull(System.getProperty("spfn.repoRoot"))
        {
            "spfn.repoRoot is not set; the module build script must pass it to the test task"
        }
    )

    private val directory: File = File(repoRoot, "Contracts/fixtures")

    /**
     * Loads a fixture through the SDK's own strict parser, so reading the evidence
     * exercises the thing being tested.
     */
    fun load(relativePath: String): SpfnCanonicalValue =
        SpfnCanonicalJson.parse(File(directory, relativePath).readBytes())

    fun bytes(relativePath: String): ByteArray = File(repoRoot, relativePath).readBytes()
}

fun SpfnCanonicalValue.members(): Map<String, SpfnCanonicalValue> =
    (this as? SpfnCanonicalValue.Obj)?.members ?: error("expected an object, got $this")

fun SpfnCanonicalValue.elements(): List<SpfnCanonicalValue> =
    (this as? SpfnCanonicalValue.Arr)?.elements ?: error("expected an array, got $this")

fun SpfnCanonicalValue.text(): String =
    (this as? SpfnCanonicalValue.Text)?.value ?: error("expected a string, got $this")

fun SpfnCanonicalValue.number(): Long =
    (this as? SpfnCanonicalValue.Integer)?.value ?: error("expected an integer, got $this")

fun Map<String, SpfnCanonicalValue>.text(key: String): String =
    (this[key] ?: error("fixture is missing '$key'")).text()

fun SpfnCanonicalValue.flag(): Boolean =
    (this as? SpfnCanonicalValue.Bool)?.value ?: error("expected a boolean, got $this")

fun Map<String, SpfnCanonicalValue>.number(key: String): Long =
    (this[key] ?: error("fixture is missing '$key'")).number()

fun Map<String, SpfnCanonicalValue>.bool(key: String): Boolean =
    (this[key] ?: error("fixture is missing '$key'")).flag()

fun Map<String, SpfnCanonicalValue>.list(key: String): List<SpfnCanonicalValue> =
    (this[key] ?: error("fixture is missing '$key'")).elements()

fun Map<String, SpfnCanonicalValue>.obj(key: String): Map<String, SpfnCanonicalValue> =
    (this[key] ?: error("fixture is missing '$key'")).members()

/** The synthetic test key the proof vectors are signed with. Never a credential. */
